Allow building using Link Time Optimizations (LTO)

keton opened the pull request:
Proposed patch set fixes "missing linker plugin" issue when building with -flto flag enabled. Also it switches default debug optimization level to Og witch saves some memory while still being compatible with debugging. Plus I've noticed that qm_rtc.c is missing implementation of qm_rtc_read(). 
I hope that you'll find proposed changes useful.

Best,
K.

keton committed:
qm_rtc.c: add implementation of qm_rtc_read()

This patch adds missing implementation of qm_rtc_read() to qm_rtc.c.
qm_rtc.h file contains declaration of this function but implementation was missing.

jeez commented on drivers/rtc/qm_rtc.c:
```diff
@@ -112,3 +112,13 @@ int qm_rtc_set_alarm(const qm_rtc_t rtc, const uint32_t alarm_val)
+{
+	QM_CHECK(rtc < QM_RTC_NUM, -EINVAL);
+	QM_CHECK(value != NULL, -EINVAL);
+	*value=QM_RTC[rtc].rtc_ccvr;
```
nitpick: fix style here, please. *value = QM_RTC[rtc].rtc_ccvr;

LGTM.

keton replied:
ok, I'll fix that

jeez commented on base.mk:
```diff
@@ -116,7 +116,7 @@ BUILD ?= release
 LDFLAGS += -Xlinker --gc-sections
 ifeq ($(BUILD), debug)
-CFLAGS += -O0 -g -DDEBUG
+CFLAGS += -Og -g -DDEBUG
```
-1 

I see your point here, but I'm afraid this has hurt us quite a few times in the past already.
-Og enables optimizations that we don't want to have in place most of the times when debugging.

What are you missing when -O0 is place more specifically, please?

keton replied:
In all my projects code size sooner or later begins to be a problem. As far as I understand -Og is almost -O2 and for other toolchains I've used produces smaller binaries. That is especially important for debug builds where you cannot use more aggressive optimizations and you have to accommodate symbols and other debug related stuff in your binary.

jeez replied:
Yes, the problem is that for the level of debugging we need in the majority of the cases, having any optimizations in place for debug builds mislead us into issues. Let's keep the default debug build as -O0 as this is something easily modifiable when one needs to change it into -Og, ok?

We could consider having a third BUILD= option that is optimized-debug or similar, but I don't think it pays off.

jeez commented:
Thanks for the contribution, @keton .

-1 due to some minor comments around.
Also, please make sure you have a look at https://github.com/quark-mcu/qmsi/blob/master/doc/commit_style.rst and follow the style there on your commit messages, ok?

thanks again!

jeez commented on base.mk:
```diff
@@ -122,7 +122,7 @@ CFLAGS += -DITA_NO_ASSERT
 endif
 else ifeq ($(BUILD), release)
-CFLAGS += -Os -fomit-frame-pointer
+CFLAGS += -Os -fomit-frame-pointer -flto
```
@dalessan this is coming as an experimental feature soon on 1.3 drop, right?
We can merge this after it 1.3 so we can have the feature validated and promoted to stable for 1.4.

dalessan replied:
Yes @jeez, exactly, in release 1.3 we will have a new experimental `BUILD=lto` build option. The plan is to have it fully validated for 1.4.

@keton, thanks a lot for your contribution. I hope it is fine for you to have this merged after 1.3.

keton replied:
If you're planning to add it anyway I think that it's ok to scrap my PR and you can work on it in your own pace. Just remember to fix missing implementation of qm_rtc_read().

jeez replied:
@keton can you please update this PR? We released a new version of QMSI last week, so a rebase to start with would be great. I still want to merge the rtc_read() and want you to check if the way we enabled LTO for now is fine or not, please.

I don't think we are going forward with the commit changing our debug builds from -O0 to -Og, however, and I hope that is not a big blocker for you. Let me know if otherwise, ok?

thanks!

jeez commented:
please rebase this PR and check my comment above.

keton commented:
Rebased on top of v1.3.0, discarded LTO and debug optimization patches and fixed code style as instructed. Hope I didn't broke anything since somehow GitHub discarded few comments under this PR.

mprinn commented:
LGTM

jeez commented on drivers/rtc/qm_rtc.c:
```diff
@@ -117,7 +117,7 @@ int qm_rtc_read(const qm_rtc_t rtc, uint32_t *const value)
 	QM_CHECK(rtc < QM_RTC_NUM, -EINVAL);
 	QM_CHECK(value != NULL, -EINVAL);
-	*value=QM_RTC[rtc].rtc_ccvr;
+	*value = QM_RTC[rtc].rtc_ccvr;
```
please just squash this into the previous commit, since you are introducing the line there.

keton replied:
@jeez I don't know how to achieve this without deleting the whole branch and making new PR. Everything I've tried leaves extra commits on top of what's already in.

jeez replied:
You can rebase your branch on top of master locally, then do a push -f to your remote branch. This will update this Pull Request. Let me know that works, ok?

jeez reviewed (changes requested):
Please check our commit msg guidelines on doc/commit_style.rst .

jeez commented:
Left 2 comments, @keton . I will merge next version of this PR.

thanks!

keton commented:
Rebased on top of 1.3.1, hopefully description matches your standards. @jeez thanks for help with git-fu. I completely forgot that I can locally delete and recreate branch.

jeez commented:
This LGTM, thanks @keton.
@danicarles @dalessan @mprinn let's merge this and cherry-pick, ok?
